# Module3Day3
## Introduction
The main functionality of this program is generation of a De Bruijn Graph (DBG) from a FASTA file of reads and a given kmersize. It will then take in a query sequence of nucleotides, iterate the solving procedure numerous times to generate many contigs, and report the longest contig containing the query sequence. It will then produce an output table containing the indices that each read contributes to the longest contig and the query sequence.


## Quick Start
To run this program can be run with default settings with the specification of a FASTA file containing reads from a sequencing experiment and a FASTA file containing a query sequence use the following command:

```bash
python main.py -reads_fasta ../data/READS.fasta -query_fasta ../data/QUERY.fasta
```

The following default output will be generated:
* ../analysis/Alleles.aln
* ../analysisAlleles.fasta
* ../analysis/DeBruijnGraph_kmer(size).json

## Installation
The required packages in this program are:

*Standard library*
* re
* sys
* os
* pickle
* json
* argparse

*Additional modules*
* [numpy](https://numpy.org/)
* [pandas](https://pandas.pydata.org/)
* [ujson](https://github.com/ultrajson/ultrajson) (Optional, for better json performance)

Additional modules can be installed with:
```bash
pip install package name
```

## Usage
This program is designed to run through the command line interface (CLI). There are a number of arguments that can be passed to configure how the program runs.

#### Standard program call
There are two required arguments to this script that must be specified in order for the program to run. This includes the `-reads_fasta` and the `-query_fasta` arguments. A typical call of this program on the command line is as such:

```bash
python main.py -reads_fasta ../data/READS.fasta -query_fasta ../data/QUERY.fasta
```

#### Help with arguments
To see all arguments, navigate to the directory containing main.py and enter:
```bash
python main.py -h
```

#### Advanced program call with more arguments
An advanced program call would include a specific kmer size, a specific output folder, the option to save a histogram CSV file of the node degrees in the graph and the read lengths, as well as increasing the number of solving iterations.
```bash
python main.py -reads_fasta READS.fasta -query_fasta QUERY.fasta -kmersize 33 -output_folder ../NewOutputFolder/ --generate_histograms -solving_iterations 100
```

#### Picking up where you left off
There is also an option to use existing kmer table, DBG, or even existing alleles that were generated by past runs. The DBG will be in .json format by default and the existing allele file will be in FASTA file format.
```bash
python main.py -existing_kmer_table ../Path/To/Kmertable.dat -existing_graph_filename ../Path/To/DBG/DBG.json -existing_allele ../Path/To/Existing/Allele/ALLELES.fasta
```

NOTE: Save a kmer table using the "--save_kmer_table" argument.

#### Turning off error correction
If you would like to skip correcting errors in the kmers that are loaded in, you can use the following argument:
```bash
python main.py -do_error_correction False
```

## Program Output
By default, the program will save its output into "../analysis/", relative to the main.py script.
The files saved by default are:
* Alleles.aln: output table of the contributions of each read to final contig and query sequence
* Alleles.fasta: [FASTA file format](https://en.wikipedia.org/wiki/FASTA_format) containing the largest constructed contig.
* DeBruijnGraph_kmer(size).json: The DBG in [JSON file format](https://www.json.org/json-en.html).

Optional output files (specified with -generate_histograms) are:
* KmerDegreeHistogram.csv: A CSV file containing the magnitude and the bin edges for a histogram of kmer degrees in the DBG.
* ReadLengthHistogram.csv: A CSV file containing the magnitude and the bin edges for a histogram of read lengths in the reads FASTA file.

#### Example output
Example output can be found in ../analysis/kmer*/. This output is from 4 different runs of the program with varying kmer sizes. Precomputed DBG are contained within these folders (example: ../analysis/kmer22/DeBruijnGraph_kmer22.json). These existing DBGs can be used to quickly run a new instance of the program without having to compute a new graph.

## Troubleshooting/Program Notes
The computational time and complexity vary depending on the kmer size that was chosen. Smaller kmer sizes (< 19) will allow the program to compute the DBG quicker (about 45 seconds to 1 min). A larger kmer size (> 22) will handle repeat sequences better than smaller kmer sizes, but will take longer to generate. Solving iterations are at a default of 50 but can be increased to larger values to sample more of the graph space and ensure the longest contig reported is in fact the best solution. This program is optimized so that it only reads in the DBG once for each program call, so increasing solving iterations will only increase the computational time by around 0.5 seconds for each additional iteration.